RPL mesh multi-node sample using qemu

pedropaulomartucci opened the pull request:
@rveerama1 as you requested, this is my setup for RPL global repair testing.

Hope it helps to reproduce the failures.

Thank you!

pfalcon commented:
@pedropaulomartucci : Thanks for the patch! As you can see, it now has conflicts with the latest master, and its build was failing. Please rebase and fix build issues.

pedropaulomartucci commented:
@pfalcon : Done.

pedropaulomartucci commented:
@pfalcon : I will take a look in the guidelines latter, probably the commit messages are wrong as you said.

rveerama1 commented:
@pedropaulomartucci : I am trying to utilize this PR to validate different RPL mesh scenarios on top of https://github.com/zephyrproject-rtos/zephyr/pull/980 and https://github.com/zephyrproject-rtos/zephyr/pull/5035 PRs. And hub.c is using 'glib'. We usually keep these files in net-tools (https://github.com/zephyrproject-rtos/net-tools). There are some changes needed, I will update you about my progress.

pedropaulomartucci commented:
@rveerama1 : I can remove the hub application from PR and move it to net-tools repository if you think is a good call, tell me.

About the RPL mesh scenarios, keep me updated. I did some tests last week with this PR on top of #980 and #5035 and noted some misbehave when pinging between nodes, even with the routes apparently right.

I will run some more tests to try reaching the root cause.

pedropaulomartucci commented:
@rveerama1 : As you suggested, I send a PR with the virtual-hub tool to the net-tools repository:
https://github.com/zephyrproject-rtos/net-tools/pull/17

The original purpose of this PR was setup RPL network scenario, which triggers a specific global-repair error. In the process, I developed the above tool, which allows setting up different network topologies for RPL using QEMU.

This way, I updated the original sample to be a guide of how to use the virtual-hub and setup multiple node RPL scenarios.

tbursztyka commented on drivers/ieee802154/ieee802154_uart_pipe.c:
```diff
@@ -24,9 +24,74 @@
+static bool received_dest_addr_matched(u8_t *rx_buffer)
+{
+	struct upipe_context *upipe = upipe_dev->driver_data;
+	struct ieee802154_context *ctx = net_if_l2_data(upipe->iface);
```
what belongs to L2, stays in L2. So you'll have to manage this differently. 

As you are implement hw filtering, L2 will take advantage of it and will send the ext addr, short addr, and pan_id to you driver: so you'll store them and use them for matching packets. No need to grab l2's context.

tbursztyka commented on drivers/ieee802154/Kconfig:
```diff
@@ -66,4 +66,48 @@ config IEEE802154_UPIPE_DRV_NAME
 	default "IEEE802154_UPIPE"
 	depends on IEEE802154_UPIPE
 
+config IEEE802154_UPIPE_PROMISCUOUS
```
Change the name here. What you have been implementing is not 15.4 promiscuous mode but "hw filtering".

See include/net/ieee802154_radio.h 

With your modifications, the upipe driver could return IEEE802154_HW_FILTER as hw capabilities.

tbursztyka commented on drivers/ieee802154/ieee802154_uart_pipe.c:
```diff
@@ -24,9 +24,74 @@
+
+	/* Check destination PAN Id */
+	if (memcmp(&rx_buffer[PAN_ID_OFFSET],
+			(u8_t *) &ctx->pan_id, PAN_ID_SIZE) != 0 &&
```
parameters are always aligned to last '(' + 1 space. 

Apply that comment everywhere

tbursztyka commented on drivers/ieee802154/ieee802154_uart_pipe.c:
```diff
@@ -24,9 +24,74 @@
+						SHORT_ADDRESS_SIZE) != 0)) {
+			return false;
+		}
+	break;
```
wrong indent here, one tab more is needed. Apply that where relevant, there are other case below.

tbursztyka commented on drivers/ieee802154/ieee802154_uart_pipe.c:
```diff
@@ -24,9 +24,74 @@
+	case DEST_ADDR_TYPE_SHORT:
+		/* First check if the destination is broadcast */
+		/* If not broadcast, check if lenght and address matches */
+		if (memcmp(&rx_buffer[DEST_ADDR_OFFSET],
```
if you want you could move that to a dedicated inline function

tbursztyka reviewed (changes requested):
Nice addition.

some style issues, but also take a look at hw filtering flag on 153.4 radio api, that will make things nicer.

pedropaulomartucci commented:
@tbursztyka : I fixed your requirements. I don't know if I did exactly what you requested about IEEE802154_HW_FILTER, tell me if any other change is required.

I had some trouble after rebasing my PR to a recent commit from master. Apparently the main problem is related with the new function: `ieee802154_enable`. Inside it, context channel is checked to not be IEEE802154_NO_CHANNEL. On the other hand, `ieee802154_init` initialized context channel value with IEEE802154_NO_CHANNEL.

For me the ideia looks ok, if we have a way to configure the channel before trying to enable the interface.

The point is, even setting up `CONFIG_NET_APP_IEEE802154_DEV_NAME="IEEE802154_UPIPE"` to trigger properly `ieee802154_settings.h` configuration function, the `ieee802154_enable` is called before it, and the channel check cause a failure when enabling the interface.

Apparently there is a tiny synchronization problem. Am I missing something?

About this PR: To validate this PR changes after rebasing I temporarily removed the IEEE802154_NO_CHANNEL check, and everything seems to work.

tbursztyka commented:
@pedropaulomartucci Sorry for the delay. About the init and enable function in L2, yes it's normal. Reason being that: channel must be properly configured externally (through net_mgmt interface) before putting the device up (via net_if_up() ). We had if up/down issues with ipv6 ND/DAD before.

Your PR seems fine to me. However, last request (looks like I forgot to ask it last time), could you split it in 2 patches: one adding hw filtering to 15.4 upipe driver, and one adding your sample? Thanks.

pedropaulomartucci commented:
@tbursztyka No problem, I will split my patch.

pedropaulomartucci committed:
drivers/ieee802154: Add hw filtering support for upipe driver

Implements HW filtering for IEEE802.15.4 UART Pipe driver.

Signed-off-by: Pedro Martucci <pedropaulomartucci@gmail.com>

pedropaulomartucci committed:
sample: net: RPL mesh sample over QEMU

Create a net sample to setup a multi-node RPL mesh network using QMEU.
To enable this, it was necessary implement a hw filter on IEEE 802.15.4
UART Pipe driver and create a QEMU pipe management on cmake.

This sample use a tool developed on zephyr net-tools repository called
virtual-hub.

Signed-off-by: Pedro Martucci <pedropaulomartucci@gmail.com>

jukkar reviewed (approved):
LGTM
